history/history-impl/src/main/java/com/duckduckgo/history/impl/HistoryRepository.kt

```diff
@@ -0,0 +1,131 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.history.impl
+
+import com.duckduckgo.common.utils.DispatcherProvider
+import com.duckduckgo.history.api.HistoryEntry
+import com.duckduckgo.history.impl.remoteconfig.HistoryFeature
+import com.duckduckgo.history.impl.store.HistoryDao
+import com.duckduckgo.history.impl.store.HistoryDataStore
+import io.reactivex.Single
+import java.time.LocalDateTime
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+
+interface HistoryRepository {
+    fun getHistoryObservable(): Single<List<HistoryEntry>>
+
+    suspend fun saveToHistory(
+        url: String,
+        title: String?,
+        query: String?,
+        isSerp: Boolean,
+    )
+
+    suspend fun clearHistory()
+
+    fun isHistoryUserEnabled(default: Boolean): Boolean
+
+    fun setHistoryUserEnabled(value: Boolean)
+
+    suspend fun clearEntriesOlderThan(dateTime: LocalDateTime)
+
+    suspend fun hasHistory(): Boolean
+}
+
+class RealHistoryRepository(
+    private val historyDao: HistoryDao,
+    private val dispatcherProvider: DispatcherProvider,
+    private val appCoroutineScope: CoroutineScope,
+    private val historyDataStore: HistoryDataStore,
+    private val historyFeature: HistoryFeature,
+) : HistoryRepository {
+
+    private var cachedHistoryEntries: List<HistoryEntry>? = null
+
+    override fun getHistoryObservable(): Single<List<HistoryEntry>> {
+        return if (cachedHistoryEntries != null) {
+            Single.just(cachedHistoryEntries)
+        } else {
+            Single.create { emitter ->
+                appCoroutineScope.launch(dispatcherProvider.io()) {
+                    try {
+                        emitter.onSuccess(fetchAndCacheHistoryEntries())
+                    } catch (e: Exception) {
+                        emitter.onError(e)
+                    }
+                }
+            }
+        }
+    }
+
+    override suspend fun saveToHistory(
+        url: String,
+        title: String?,
+        query: String?,
+        isSerp: Boolean,
+    ) {
+        withContext(dispatcherProvider.io()) {
+            historyDao.updateOrInsertVisit(
+                url,
+                title ?: "",
+                query,
+                isSerp,
+                LocalDateTime.now(),
+            )
+            fetchAndCacheHistoryEntries()
+        }
+    }
+
+    override suspend fun clearHistory() {
+        withContext(dispatcherProvider.io()) {
+            cachedHistoryEntries = null
+            historyDao.deleteAll()
```

Comment:
On a debug build, with 30k visits, this delete is taking 25 seconds to complete :\

I don't expect that number on a release build, but it's going to impact the fire process time. Let's take a look at this one.

Reply:
I have a PoC for you to check.

```
    @Query("DELETE FROM visits_list")
    suspend fun deleteVisits()

    @Query("DELETE FROM history_entries")
    fun deleteEntries()

    @Transaction
    suspend fun deleteAll() {
        deleteVisits()
        deleteEntries()
    }
```

This seems to fix the long times.